File: 0051/solution.java
import java.util.*;

class Solution {
    Set<Integer> cols = new HashSet<>();
    Set<Integer> diagonalsLeft = new HashSet<>();
    Set<Integer> diagonalsRight = new HashSet<>();
    List<List<String>> results = new ArrayList<>();
    List<String> currentBoard = new ArrayList<>();

    public List<List<String>> solveNQueens(int n) {
        backtrack(0, n);
        return results;
    }

    private void backtrack(int row, int n) {
        if (row == n) {
            results.add(new ArrayList<>(currentBoard));
            return;
        }
        for (int col = 0; col < n; col++) {
            if (cols.contains(col) || diagonalsLeft.contains(row - col) || diagonalsRight.contains(row + col)) {
                continue;
            }
            char[] rowArray = new char[n];
            Arrays.fill(rowArray, '.');
            rowArray[col] = 'Q';
            String rowString = new String(rowArray);

            currentBoard.add(rowString);
            cols.add(col);
            diagonalsLeft.add(row - col);
            diagonalsRight.add(row + col);

            backtrack(row + 1, n);

            currentBoard.remove(currentBoard.size() - 1);
            cols.remove(col);
            diagonalsLeft.remove(row - col);
            diagonalsRight.remove(row + col);
        }
    }
}

File: 0051/solution.py
class Solution:
    def solveNQueens(self, n: int) -> list[list[str]]:
        cols = set()
        left_diag = set()
        right_diag = set()
        res = []
        curr = []

        def backtrack(row):
            if row == n:
                res.append(curr.copy())
                return
            for col in range(n):
                if col in cols or col - row in left_diag or col + row in right_diag:
                    continue
                cols.add(col)
                left_diag.add(col - row)
                right_diag.add(col + row)
                curr.append("." * col + "Q" + "." * (n - col - 1))
                backtrack(row + 1)
                curr.pop()
                cols.remove(col)
                left_diag.remove(col - row)
                right_diag.remove(col + row)

        backtrack(0)
        return res
